woocommerce/woocommerce
# [Accessibility] Improve checkout error messages

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This is the second PR dedicated to close the #37602 issue.
**First PR**: https://github.com/woocommerce/woocommerce/pull/48738

Closes #37602  .



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This PR improves the form error messages on the classic checkout page.

1. Add a product to the cart.
2. Go to the classic checkout page.
3. Turn on the screen reader.
4. Submit the form with the required fields empty.
6. Verify a notice element appears at the top of the form. The notice element style may vary per theme.
![image](https://github.com/woocommerce/woocommerce/assets/9420947/ab723a51-84b8-4126-aa68-b597ce1130e0)
7. Verify the notice element contains an error message for each one of the required fields with error.
8. Verify the focus went to the notice element and the screen reader announced its content.
9. Verify you can tab through the error messages inside the notice element.
10. The hover style of the links inside the notice element depends on the theme installed.
11. Verify each one of the error messages moves the focus to the associated form field once clicked.
14. Verify the fields with errors have an inline message below them.
15. By default, the style of the inline message is: `color: #a00;`,  `font-size: .75em;`, and `margin-bottom: 0;`. However, it can vary depending on the installed theme.
![image](https://github.com/woocommerce/woocommerce/assets/9420947/8908cb9c-6ae7-46e5-a4f1-23096d649188)
16. The inline messages should disappear once the field value is updated and not appear again until the form has been submitted. This behavior is because the error messages come translated from the server and can be different depending on the value of the input.
17. Perform the steps above logged in/out.
18. Perform the test above with the "Shipping to a different address" form opened.
19. The logic that validates the inputs on the client side when their value change remains the same.
20. Perform the tests against the WP default themes:
    - Twenty Seventeen
    - Twenty Nineteen
    - Twenty Twenty
    - Twenty Twenty-One
    - Twenty Twenty-Two
    - Twenty Twenty-Three
    - Twenty Twenty-Four



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
